src/Microsoft.AspNetCore.Razor.Language/Syntax/SyntaxNodeExtensions.cs

```diff
@@ -1,13 +1,339 @@
+        }
+
+        /// <summary>
+        /// Creates a new tree of nodes with specified old node replaced with a new nodes.
+        /// </summary>
+        /// <typeparam name="TRoot">The type of the root node.</typeparam>
+        /// <param name="root">The root of the tree of nodes.</param>
+        /// <param name="oldNode">The node to be replaced; a descendant of the root node and an element of a list member.</param>
+        /// <param name="newNodes">A sequence of nodes to use in the tree in place of the old node.</param>
+        public static TRoot ReplaceNode<TRoot>(this TRoot root, SyntaxNode oldNode, IEnumerable<SyntaxNode> newNodes)
+            where TRoot : SyntaxNode
+        {
+            return (TRoot)root.ReplaceNodeInListCore(oldNode, newNodes);
+        }
+
+        /// <summary>
+        /// Creates a new tree of nodes with new nodes inserted before the specified node.
+        /// </summary>
+        /// <typeparam name="TRoot">The type of the root node.</typeparam>
+        /// <param name="root">The root of the tree of nodes.</param>
+        /// <param name="nodeInList">The node to insert before; a descendant of the root node an element of a list member.</param>
+        /// <param name="newNodes">A sequence of nodes to insert into the tree immediately before the specified node.</param>
+        public static TRoot InsertNodesBefore<TRoot>(this TRoot root, SyntaxNode nodeInList, IEnumerable<SyntaxNode> newNodes)
+            where TRoot : SyntaxNode
+        {
+            return (TRoot)root.InsertNodesInListCore(nodeInList, newNodes, insertBefore: true);
+        }
+
+        /// <summary>
+        /// Creates a new tree of nodes with new nodes inserted after the specified node.
+        /// </summary>
+        /// <typeparam name="TRoot">The type of the root node.</typeparam>
+        /// <param name="root">The root of the tree of nodes.</param>
+        /// <param name="nodeInList">The node to insert after; a descendant of the root node an element of a list member.</param>
+        /// <param name="newNodes">A sequence of nodes to insert into the tree immediately after the specified node.</param>
+        public static TRoot InsertNodesAfter<TRoot>(this TRoot root, SyntaxNode nodeInList, IEnumerable<SyntaxNode> newNodes)
+            where TRoot : SyntaxNode
+        {
+            return (TRoot)root.InsertNodesInListCore(nodeInList, newNodes, insertBefore: false);
+        }
+
+        public static string GetContent<TNode>(this TNode node) where TNode : SyntaxNode
+        {
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            var tokens = node.DescendantNodes().Where(n => n.IsToken).Cast<SyntaxToken>();
+            var content = string.Concat(tokens.Select(t => t.Content));
+            return content;
+        }
+
+        public static string GetTagName(this MarkupTagBlockSyntax tagBlock)
```

Comment:
These are currently extension methods. But these will go away once the structure of a Tag Block is improved.